c/cobuild.c

```diff
@@ -0,0 +1,1043 @@
+  if (err) {
+    printf("callback failed: %d", err);
+    // terminated immediately
+    ckb_exit(err);
+  }
+exit:
+  return err;
+}
+
+int ckb_generate_otx_smh(mol2_cursor_t message_cursor, uint8_t *smh,
+                         const OtxStart *start, const Otx *size) {
+  int err = 0;
+  blake2b_state ctx;
+  size_t count = 0;
+  new_otx_blake2b(&ctx);
+  printf(
+      "start_input_cell = %d, start_output_cell = %d, start_cell_deps = %d, "
+      "start_header_deps = %d",
+      start->start_input_cell, start->start_output_cell, start->start_cell_deps,
+      start->start_header_deps);
+  printf(
+      "input_cells = %d, output_cells = %d, cell_deps = %d, header_deps = %d",
+      size->input_cells, size->output_cells, size->cell_deps,
+      size->header_deps);
+
+  err = ckb_hash_cursor(&ctx, message_cursor);
+  CHECK(err);
+  count += message_cursor.size;
+
+  BLAKE2B_UPDATE(&ctx, &size->input_cells, 4);
+  count += 4;
+
+  // hash input cell and data
+  for (size_t index = start->start_input_cell;
+       index < (start->start_input_cell + size->input_cells); index++) {
+    // CellInput
+    uint8_t input[128];
+    uint64_t input_len = sizeof(input);
+    err = ckb_load_input(input, &input_len, 0, index, CKB_SOURCE_INPUT);
+    CHECK(err);
+    BLAKE2B_UPDATE(&ctx, input, input_len);
+    count += input_len;
+
+    err = hash_cell(&ctx, index, CKB_SOURCE_INPUT, &count);
+    CHECK(err);
+  }
+  // hash output cell and data
+  BLAKE2B_UPDATE(&ctx, &size->output_cells, 4);
+  count += 4;
+  for (size_t index = start->start_output_cell;
+       index < (start->start_output_cell + size->output_cells); index++) {
+    err = hash_cell(&ctx, index, CKB_SOURCE_OUTPUT, &count);
+    CHECK(err);
+  }
+
+  // hash cell deps
+  BLAKE2B_UPDATE(&ctx, &size->cell_deps, 4);
+  count += 4;
+  err = hash_cell_deps(&ctx, &count, start->start_cell_deps, size->cell_deps);
+  CHECK(err);
+
+  // hash header deps
+  BLAKE2B_UPDATE(&ctx, &size->header_deps, 4);
+  count += 4;
+  for (size_t index = start->start_header_deps;
+       index < (start->start_header_deps + size->header_deps); index++) {
+    uint8_t header_dep[32];
+    uint64_t header_dep_len = sizeof(header_dep);
+    err = ckb_load_header(header_dep, &header_dep_len, 0, index,
+                          CKB_SOURCE_HEADER_DEP);
+    CHECK(err);
+    BLAKE2B_UPDATE(&ctx, header_dep, header_dep_len);
+    count += header_dep_len;
+  }
+  printf("ckb_generate_otx_smh totally hashed %d bytes", count);
+  blake2b_final(&ctx, smh, BLAKE2B_BLOCK_SIZE);
+exit:
+  return err;
+}
+
+static int get_witness_layout(size_t index,
+                              WitnessLayoutId *witness_layout_id) {
+  uint32_t id = 0;
+  uint64_t id_len = sizeof(id);
+  int err = ckb_load_witness(&id, &id_len, 0, index, CKB_SOURCE_INPUT);
+  if (err) return err;
+  if (id_len < sizeof(id)) {
+    return ERROR_GENERAL;
+  }
+  if (id == WitnessLayoutSighashAll || id == WitnessLayoutSighashAllOnly ||
```

Comment:
This is due to the limitation of molecule-c2. If the data structure is invalid, it may quit(panic) early. 
Note, it's hard to make it panic free to parse an invalid data. e.g. overflow

Reply:
I was confused: if a witness data structure is invalid, the smart contract should abort anyway. What's the problem of quiting / panicing / aborting here?

Reply:
Actually the "get_witness_layout" should be renamed to "detect_witness_layout" here. It's possible that the data structure is WitnessArgs or invalid.

Reply:
Let's try this again:

* Make a WitnessLayout data structure from a cursor.
* Run a verifier to ensure the correctness of the WitnessLayout data structure. A verifier here should be panic free, regardless of what data is passed here.
* If the verifier passes, you have a correct data structure, no panic will happen afterwards
* If the verifier fails, you try the next one

Reply:
Currently, molecule-c2 is not panic free.

Reply:
I'm not talking about molecule-c2, which still does not have a verifier. I'm talking about potential verifier needed here. molecule-c2's most current accessor code can stay as it is with the potential of panicing at times, as long as we can provide verifier code that is panic free(and should be panic free, we have already done similar thing in Rust)

Reply:
Panic free is not very easy. Every operator (+, -) should be checked to avoid overflowing.

Reply:
This is not about easy or hard. You are merely talking about efforts and paying attention to the code when you write it. If we don't have a solution to a thing, I would call it hard. Here it is just about efforts.

I can experiment on it if you guys find it hard to do.